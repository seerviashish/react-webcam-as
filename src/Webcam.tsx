import React, { useEffect, useRef } from 'react';

const EXIF = require('exif-js');

interface IWebcam {
  accept?: string;
  className?: string;
  iconComponent?: any;
  label: string;
  onChange: (data: Data) => any;
  style?: { button?: React.CSSProperties; input?: React.CSSProperties };
}

type Data = {
  dataUrl: string | ArrayBuffer | null;
  dimension?: DimensionType;
  type: string;
};

type DimensionType = {
  h: number;
  w: number;
};

const Webcam: React.FC<IWebcam> = ({
  accept,
  className,
  iconComponent,
  label,
  onChange,
  style,
}) => {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (!inputRef?.current) return;
    inputRef.current.addEventListener('change', cameraEventListener);
    // eslint-disable-next-line consistent-return
    return () =>
      inputRef.current?.removeEventListener('change', cameraEventListener);
  }, [inputRef]);

  const cameraEventListener = async (event: any) => {
    await processFile(event.target.files);
  };

  const handleButton = (event: React.MouseEvent<HTMLButtonElement>): void => {
    event.preventDefault();
    const inputElement = inputRef.current;
    if (inputElement) {
      inputElement.click();
    }
  };

  const getImageDimensions = (file: string): Promise<DimensionType> =>
    new Promise((resolve, reject) => {
      const i = new Image();
      i.onload = () => resolve({ w: i.width, h: i.height });
      i.onerror = () => reject(new Error('E01: Get image dimensions failed'));
      i.src = file;
    });

  const createImage = (
    data: string | ArrayBuffer | null
  ): Promise<HTMLImageElement> =>
    new Promise((resolve, reject) => {
      if (data == null) {
        reject(new Error('E02: Create image data null'));
      }
      const img = document.createElement('img');
      img.onload = () => resolve(img);
      img.onerror = () => reject(new Error('E03: Image creation failed'));
      img.src = data as string;
    });

  const fixOrientation = async (img: HTMLImageElement): Promise<string> => {
    const { width } = img;
    const { height } = img;
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');
    if (!ctx) {
      throw new Error('E04: Canvas context is null');
    }
    let exifError = false;
    try {
      await EXIF.getData(img, () => {
        const orientation = EXIF.getAllTags(this).Orientation;
        if (orientation > 4 && orientation < 9) {
          canvas.width = height;
          canvas.height = width;
        } else {
          canvas.width = width;
          canvas.height = height;
        }
        switch (orientation) {
          case 2:
            ctx.transform(-1, 0, 0, 1, width, 0);
            break;
          case 3:
            ctx.transform(-1, 0, 0, -1, width, height);
            break;
          case 4:
            ctx.transform(1, 0, 0, -1, 0, height);
            break;
          case 5:
            ctx.transform(0, 1, 1, 0, 0, 0);
            break;
          case 6:
            ctx.transform(0, 1, -1, 0, height, 0);
            break;
          case 7:
            ctx.transform(0, -1, -1, 0, height, width);
            break;
          case 8:
            ctx.transform(0, -1, 1, 0, 0, width);
            break;
          default:
            break;
        }
        ctx.drawImage(img, 0, 0);
      });
    } catch (e) {
      exifError = true;
      // eslint-disable-next-line no-console
      console.error(e);
    }
    return exifError ? img.src : canvas.toDataURL('image/jpeg');
  };

  const processFile = async (files: FileList) => {
    if (files.length === 0) {
      throw new Error('E05: File not selected');
    }
    const file = files[0];
    const contentBuffer = await readFileAsync(file);
    let fileDimension = null;
    let fileFixed: string | null = null;
    let fileDataResponse = null;
    if (accept?.includes('image') || !accept) {
      const imageFileElement: HTMLImageElement = await createImage(
        contentBuffer
      );
      fileFixed = await fixOrientation(imageFileElement);
      fileDimension = await getImageDimensions(fileFixed);
      fileDataResponse = {
        type: accept || 'image/*',
        dataUrl: fileFixed,
        dimension: fileDimension,
      };
    } else {
      fileDataResponse = {
        type: accept,
        dataUrl: contentBuffer,
      };
    }
    onChange(fileDataResponse);
  };

  const readFileAsync = async (
    file: File
  ): Promise<string | ArrayBuffer | null> =>
    new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => {
        resolve(reader.result);
      };
      reader.onerror = reject;
      reader.readAsDataURL(file);
    });

  return (
    <div>
      <input
        ref={inputRef}
        type="file"
        accept={accept || 'image/*'}
        style={{ display: 'none' }}
        onClick={(event: any) => {
          // eslint-disable-next-line no-param-reassign
          event.target.value = null;
        }}
      />
      <button
        className={className}
        style={style?.button}
        onClick={handleButton}
        type="submit"
      >
        {iconComponent}
        {label}
      </button>
    </div>
  );
};

export default Webcam;
