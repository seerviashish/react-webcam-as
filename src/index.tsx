/**
 * @class WebcamNative
 */

import * as React from "react";
const EXIF = require("exif-js");
import styles from "./styles.css";

export type Props = {
  accept?: string;
  onChange: (data: Object) => any;
  buttonStyle?: Object;
  label: string;
  iconComponent?: any;
};

export default class WebcamNative extends React.Component<Props> {
  private myInput = React.createRef<any>();
  constructor(props: Readonly<Props>) {
    super(props);
  }

  componentDidMount() {
    this.myInput.current.addEventListener(
      "change",
      this.cameraEventListener,
      this.myInput
    );
  }

  componentWillUnmount() {
    this.myInput.current.removeEventListener(
      "change",
      this.cameraEventListener,
      this.myInput
    );
  }

  private cameraEventListener = async (event: any) => {
    try {
      await this.processFile(event.target.files);
    } catch (error) {
      console.log(error);
    }
  };

  private handleButton = (event: any) => {
    event.preventDefault();
    const node = this.myInput.current;
    if (node) {
      node.click();
    }
  };

  private getImageDimensions = (file: any) => {
    return new Promise(function(resolved) {
      var i = new Image();
      i.onload = function() {
        resolved({ w: i.width, h: i.height });
      };
      i.src = file;
    });
  };

  private createImage = (data: any) => {
    return new Promise(resolve => {
      const img = document.createElement("img");
      img.onload = () => resolve(img);
      img.src = data;
    });
  };

  private fixOrientation = async (img: any) => {
    let width = img.width,
      height = img.height,
      canvas = document.createElement("canvas"),
      ctx = canvas.getContext("2d");
    await EXIF.getData(img, () => {
      const orientation = EXIF.getAllTags(this).Orientation;
      if (4 < orientation && orientation < 9) {
        canvas.width = height;
        canvas.height = width;
      } else {
        canvas.width = width;
        canvas.height = height;
      }
      switch (orientation) {
        case 2:
          ctx && ctx.transform(-1, 0, 0, 1, width, 0);
          break;
        case 3:
          ctx && ctx.transform(-1, 0, 0, -1, width, height);
          break;
        case 4:
          ctx && ctx.transform(1, 0, 0, -1, 0, height);
          break;
        case 5:
          ctx && ctx.transform(0, 1, 1, 0, 0, 0);
          break;
        case 6:
          ctx && ctx.transform(0, 1, -1, 0, height, 0);
          break;
        case 7:
          ctx && ctx.transform(0, -1, -1, 0, height, width);
          break;
        case 8:
          ctx && ctx.transform(0, -1, 1, 0, 0, width);
          break;
        default:
          break;
      }
      ctx && ctx.drawImage(img, 0, 0);
    });
    return canvas.toDataURL("image/jpeg");
  };

  private processFile = async (files: FileList) => {
    const { accept } = this.props;
    try {
      if (files.length === 0) {
        throw new Error("Files not selected");
      }
      let file = files[0];
      let contentBuffer = await this.readFileAsync(file);
      let filedimension = null;
      let fileFixed = null;
      let fileDataResponse = null;
      if ((accept && accept.includes("image")) || !accept) {
        const fileData = await this.createImage(contentBuffer);
        fileFixed = await this.fixOrientation(fileData);
        filedimension = await this.getImageDimensions(fileFixed);
        fileDataResponse = {
          type: accept ? accept : "image/*",
          dataUrl: fileFixed,
          dimension: filedimension
        };
      } else {
        fileDataResponse = {
          type: accept,
          dataUrl: contentBuffer
        };
      }
      this.props.onChange(fileDataResponse);
    } catch (error) {
      console.log(error);
    }
  };

  private readFileAsync = async (file: File) => {
    return new Promise((resolve, reject) => {
      let reader = new FileReader();
      reader.onload = () => {
        resolve(reader.result);
      };
      reader.onerror = reject;
      reader.readAsDataURL(file);
    });
  };

  render() {
    const { accept, buttonStyle, label, iconComponent } = this.props;

    return (
      <div>
        <input
          ref={this.myInput}
          type="file"
          accept={accept ? accept : "image/*"}
          style={{ display: "none" }}
          onClick={(event: any) => {
            event.target.value = null;
          }}
        />
        <button
          className={styles.ripple}
          style={buttonStyle}
          onClick={this.handleButton}
        >
          {iconComponent}
          {label}
        </button>
      </div>
    );
  }
}
