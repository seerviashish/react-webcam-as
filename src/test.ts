import WebcamNative from "./";

describe("ExampleComponent", () => {
  it("is truthy", () => {
    expect(WebcamNative).toBeTruthy();
  });
});
