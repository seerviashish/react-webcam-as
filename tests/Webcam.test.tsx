import renderer from 'react-test-renderer';
import { expect, it, vi } from 'vitest';
import { Webcam } from '../src';

it('render webcam correctly', () => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const handleOnChange = vi.fn().mockImplementation((data) => 0);
  const tree = renderer
    .create(
      <Webcam
        accept="image/*"
        label="Camera"
        onChange={handleOnChange}
        style={{ button: { textTransform: 'none' } }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
