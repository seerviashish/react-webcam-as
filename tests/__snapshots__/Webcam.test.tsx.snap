// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render webcam correctly 1`] = `
<div>
  <input
    accept="image/*"
    onClick={[Function]}
    style={
      {
        "display": "none",
      }
    }
    type="file"
  />
  <button
    onClick={[Function]}
    style={
      {
        "textTransform": "none",
      }
    }
    type="submit"
  >
    Camera
  </button>
</div>
`;
